lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -497,25 +443,40 @@ def get_kv_requests(self, relation_id: int | None = None) -> List[KVRequest]:
                 )
         return kv_requests
 
-    def _credentials_issued_for_request(self, nonce: str, relation_id: int) -> bool:
-        """Return whether credentials have been issued for the request."""
-        relation = self.model.get_relation(self.relation_name, relation_id)
-        if not relation:
-            return False
+    def set_kv_data(
+        self,
+        relation: ops.Relation,
+        mount: str,
+        ca_certificate: str,
+        vault_url: str,
+        nonce: str,
+        secret_id: str,
+    ):
+        """Set the kv data on the relation."""
         credentials = self.get_credentials(relation)
-        return credentials.get(nonce) is not None
+        credentials[nonce] = secret_id
+        try:
+            self.juju_facade.set_app_relation_data(
+                relation_name=self.relation_name,
+                relation=relation,
+                data={
+                    "mount": mount,
+                    "ca_certificate": ca_certificate,
+                    "vault_url": vault_url,
+                    "credentials": json.dumps(credentials, sort_keys=True),
+                },
+            )
+        except NotLeaderError:
+            return
```

Comment:
In some places in this file we have

```python
if not self.juju_facade_is_leader:
    return
do_x()
```

And in other places, like here, we have:

```python
try:
    self.do_x()
except NotLeaderError:
    return
```

Let's stick with one of the approaches

Reply:
Good point, there is no need to have both in this file.